src/gps.c

```diff
@@ -0,0 +1,223 @@
+          {
+            gps_data_read.vdop = gps_data.dop.vdop;
+          }
+          if (isnan(gps_data.dop.hdop) == 0)
+          {
+            gps_data_read.hdop = gps_data.dop.hdop;
+          }
+
+          // Sat in view:
+          if ((gps_data.set & LATLON_SET))
+          {
+            gps_data_read.satellites = gps_data.satellites_used;
+          }
+
+          pthread_mutex_unlock (&data_mutex);
+        }
+      }
+    }
+  }
+
+  gps_stream(&gps_data, WATCH_DISABLE, NULL);
+  gps_close(&gps_data);
+
+  pthread_exit ((void *)0);
+}
+
+
+/**
+ * Submit the data.
+ */
+static void gps_collectd_submit (const char *type, gauge_t value)
+{
+  value_t values[1];
+  value_list_t vl = VALUE_LIST_INIT;
+
+  values[0].gauge = value;
+
+  vl.values = values;
+  vl.values_len = 1;
+  sstrncpy (vl.host, hostname_g, sizeof (vl.host));
+  sstrncpy (vl.plugin, "gps", sizeof (vl.plugin));
+  sstrncpy (vl.type, type, sizeof (vl.type));
+  sstrncpy (vl.type_instance, "gps", sizeof (vl.type_instance));
+
+  plugin_dispatch_values (&vl);
+}
+
+
+/**
+ * Read the data and submit.
+ */
+static int gps_collectd_read ()
+{
+  pthread_mutex_lock (&data_mutex);
+  gps_collectd_submit("gps_hdop", (gauge_t) gps_data_read.hdop);
+  gps_collectd_submit("gps_vdop", (gauge_t) gps_data_read.vdop);
+  gps_collectd_submit("gps_sat", (gauge_t) gps_data_read.satellites);
+  printf ("gps: hdop=%1.3f, vdop=%1.3f, sat=%02d.\n", 
+    gps_data_read.hdop,
+    gps_data_read.vdop,
+    gps_data_read.satellites
+  );
+  pthread_mutex_unlock (&data_mutex);
+  return (0);
+}
+
+
+/**
+ * Read configuration.
+ */
+static int gps_collectd_config (const char *key, const char *value)
+{
+  if (strcasecmp (key, "Host") == 0) {
+    if (gps_data_config.host != NULL) free (gps_data_config.host);
+      gps_data_config.host = sstrdup (value);
+  }
+  if (strcasecmp (key, "Port") == 0) {
+    if (gps_data_config.port != NULL) free (gps_data_config.port);
+      gps_data_config.port = sstrdup (value);
+  }
+  if (strcasecmp (key, "Timeout") == 0) {
+     gps_data_config.timeout = (int) strtol (value, NULL, 1000);
+  }
+  return (0);
+}
+
+
+/**
+ * Init.
+ */
+static int gps_collectd_init (void)
+{
+  int err = 0;
+
+  printf ("gps: will use %s:%s with timeout %d.\n", gps_data_config.host, gps_data_config.port, gps_data_config.timeout);
+
+  err = plugin_thread_create (&connector, NULL, gps_collectd_thread, NULL);
+
+  if (err != 0) {
+    WARNING ("pthread_create() failed.");
+    return (-1);
+  }
+
+  return (0);
+}
+
+
+/**
+ * Shutdown.
+ */
+static int gps_collectd_shutdown (void)
+{
+  if (connector != ((pthread_t) 0)) {
+    pthread_kill (connector, SIGTERM);
+    connector = (pthread_t) 0;
+  }
+
+  sfree (gps_data_config.port);
+  sfree (gps_data_config.host);
+
+  return (0);
+}
+
+/**
+ * Register the module.
+ */
+void module_register (void)                                                                    
+{
+  gps_data_config.host = sstrdup ("localhost");
```

Comment:
Please add the default values into a define, i.e. something like this near the top of the file:
```c
#define GPS_DEFAULT_NODE "localhost"
#define GPS_DEFAULT_SERVICE "2947"
```